### PR TITLE
Stop ignoring the change of verbosity when using Gc.set

### LINKED ISSUE #11692: GC doesn't print verbose output with OCaml 5
The following program is very verbose on 4.14, but produces no output at all in the stderr on OCaml 5:

```ocaml
let () =
  Gc.set { (Gc.get ()) with verbose = 0x017 };
  for _ = 0 to 1000 do
    let s = String.make 1000 'a' in
    print_endline s
  done
```

### PR DESCRIPTION
**Pull request by abbysmal**
This commit re-introduces the global variable `caml_verb_gc` as an atomic, addressing #11692.

The changes to it are currently ignored because the code to do it is missing, and the `verb_gc` flag is located in `caml_params`, which is immutable after initialisation. 

The implementation requires the re-introduction of a global variable (`caml_verb_gc`), this time as a global atomic.

To note: I decided to split the streamlining of log levels in a separate PR, because this is actually a fair amount of work and reading. (see https://github.com/ocaml/ocaml/issues/11692#issuecomment-1301599862 for details.)

### CODE REVIEW

File: runtime/gc_ctrl.c [hunk 1]
```diff
@@ -171,6 +172,10 @@ CAMLprim value caml_gc_set(value v)
                      ARCH_INTNAT_PRINTF_FORMAT "u%%\n", caml_percent_free);
   }
 
+  newverb_gc = Long_val (Field (v, 3));
+
+  atomic_store_rel(&caml_verb_gc, newverb_gc);
```
Comment (gasche) [inline runtime/gc_ctrl.c]: Wait, do you mean `atomic_store_rel(ease)` or `atomic_store_relaxed`? (Currently you are pairing store-releases with read-relaxed, which I'm not sure makes sense.)

Note: I find this naming choice confusing and I wish we used `_release` and `_acquire` instead of `_rel` and `_acq`.

**Review by gasche — APPROVED** [review]
I reviewed the change and believe it is correct. It fixes a small issue.

(See minor comment on the atomic mode of writing the value.)

Do I correctly understand, while re-reading the coe of `caml_gc_set`, that most of the other global values set by this function (`caml_custom_{minor,major}_ratio` and `caml_percent_free`) are not declared as atomics and therefore their access is currently completely racy?

(Shouldn't we make them relaxed atomic, as you are doing here?)

File: runtime/gc_ctrl.c (hunk 1 above)
Comment (abbysmal) [reply runtime/gc_ctrl.c]: You are perfectly right, good catch.
I agree that the naming is quite confusing, but I wonder how doable it would be to do a full renaming of the codebase without annoying everyone too much.

Comment (gadmm) [reply runtime/gc_ctrl.c]: `release` is already abbreviated into `rel` in `memory_order_acq_rel` so that's a good reminder.

File: runtime/gc_ctrl.c [hunk 2]
```diff
@@ -129,7 +129,7 @@ CAMLprim value caml_gc_get(value v)
   res = caml_alloc_tuple (11);
   Store_field (res, 0, Val_long (Caml_state->minor_heap_wsz));  /* s */
   Store_field (res, 2, Val_long (caml_percent_free));           /* o */
-  Store_field (res, 3, Val_long (caml_params->verb_gc));        /* v */
+  Store_field (res, 3, Val_long (caml_verb_gc));        /* v */
```
Comment (kayceesrk) [inline runtime/gc_ctrl.c]: `Val_long (caml_verb_gc)` does a non-atomic read. 

```suggestion
  Store_field (res, 3, Val_long (atomic_load_relaxed(&caml_verb_gc)));        /* v */
```

The indentation of the comment needs to be fixed.

File: runtime/gc_ctrl.c [hunk 3]
```diff
@@ -171,6 +172,10 @@ CAMLprim value caml_gc_set(value v)
                      ARCH_INTNAT_PRINTF_FORMAT "u%%\n", caml_percent_free);
   }
 
+  newverb_gc = Long_val (Field (v, 3));
+
+  atomic_store_relaxed(&caml_verb_gc, newverb_gc);
```
Comment (kayceesrk) [inline runtime/gc_ctrl.c]:
```suggestion
  atomic_store_relaxed(&caml_verb_gc, Long_val (Field (v, 3)));
```
and avoid the new variable declaration.

File: runtime/minor_gc.c [hunk 4]
```diff
@@ -657,12 +657,12 @@ void caml_empty_minor_heap_promote(caml_domain_state* domain,
 }
 
 void caml_do_opportunistic_major_slice
-  (caml_domain_state* domain_unused, void* unused)
+  (caml_domain_state* domain, void* unused)
```
Comment (kayceesrk) [inline runtime/minor_gc.c]: Nit: What is the reasoning behind this renaming? `domain_unused` name indicated that the variable was unused. It still remains unused in this PR. FWIW, there are other `domain_unused` parameters in the same file. If the plan is to change all `domain_unused` to `domain` it would be useful to do it wholesale and in a separate PR.

**Comment by gasche** [comment]
A naive question while looking at the code again: I would expect the value of `caml_params->verb_gc` (which is still declared after this PR) to inform the default value of `caml_verb_gc` at program-initialization time. But I don't see new logic to do this in the code. How does it happen?
(Or: `startup_aux.c` could initialize `caml_verb_gc` instead of `params->verb_gc`, as is donee for `caml_runtime_warnings`.)

**Comment by abbysmal** [comment]
> A naive question while looking at the code again: I would expect the value of caml_params->verb_gc (which is still declared after this PR) to inform the default value of caml_verb_gc at program-initialization time. But I don't see new logic to do this in the code. How does it happen?

It is currently not even working.
I am sorry for wasting reviewers time on something so simple, I will fix this mess and do better next time, these are too many silly mistakes.

I think we can replace the `caml_params` for a non-atomic write to `caml_verb_gc` (since `scanmult` only does non-atomic writing operations). It should be safe because there can't be other domains or threads when we are parsing runparams.

My apologies again.

**Comment by gasche** [comment]
No opinion on whether we want to keep a record of the initial value in `caml_params` or not. It is important in some cases (for per-domain value, this is the value we are currently using on startup of a new domain), not in this one. You could get rid of the field for simplicity, or keep it for consistency, both are very reasonable.

> I think we can replace the caml_params for a non-atomic write to caml_verb_gc (since scanmult only does non-atomic writing operations). It should be safe because there can't be other domains or threads when we are parsing runparams.

I'm not sure what you mean by `non-atomic write`; I would prefer to see a relaxed write than a direct assignment (which I'm not sure how to read on atomic types). Another option would be to use strongly consistent operations in all the cases where performance is not a concern at all. In any case I agree with your reasoning that no specific synchronization is needed there.

**Comment by abbysmal** [comment]
Hello,

Sorry, I was rather busy last week and I forgot to push a commit that addresses the outstanding issues with the branch.

I went with your initial suggestion of removing `verb_gc` from `caml_params` and make the code handling these update `caml_verb_gc` instead.

This should fix the remaining issues, sorry for the delay. :)

File: runtime/gc_ctrl.c [hunk 5]
```diff
@@ -171,6 +171,8 @@ CAMLprim value caml_gc_set(value v)
                      ARCH_INTNAT_PRINTF_FORMAT "u%%\n", caml_percent_free);
   }
 
+  atomic_store_relaxed(&caml_verb_gc, Long_val (Field (v, 3)));
```
Comment (gasche) [inline runtime/gc_ctrl.c:174]: Nitpick: in this function it is not easy to check that we are in fact taking all the parameters into account. I think that one change that would improve readability in this respect is to make all the `Long_val (Field (v, n))` read at the beginning of the function, giving names to the results, in the same order as the fields are stored in the `caml_gc_get` function. Then it would be very easy to check that they both use the same fields (0, 2, 3, 5, 8, 9, 10).

**Review by gasche — APPROVED** [review]
I reviewed the PR and believe that the code is correct. (See minor nitpick comment inline.) Thanks!

**Comment by gasche** [comment]
As a non-memory-model-expert, it is not obvious to me what memory order we should be using here -- in the PR, we always use `relaxed`. My understanding is that a few reads are in fact performance-sensitive (in particular the one in the logging function for the GC, which is used a lot in the runtime), but most places are not. (So one could think of using a sequential order in all non-sensitive places). But I believe that using `relaxed` all over the place gives the following behavior which is actually the one we want (and not try to give stronger guarantees):

- setting changes in one place may be seen by other domains at an arbitrary/unspecified point, but *no later* than the next synchronization between them (due to runtime code or to user synchronization)
- in particular, setting changes at initialization time are visible to all domains at startup,
- and for setting changes caused by user code, all domains are guaranteed to see the change after the next minor GC

This seems completely fine, arguably the right specification, for a global variable that is performance-sensitive in the runtime code, and for which the promptness of the change propagation (in multi-domain scenarios) is not important for correctness of the program.

(I wonder if we should consider forcing that `Gc.set` performs a STW synchronization at the end to guarantee that the changes are always propagated immediately. We already have to perform a STW synchronization when increasing the minor heap size.)

**Comment by abbysmal** [comment]
> (I wonder if we should consider forcing that Gc.set performs a STW synchronization at the end to guarantee that the changes are always propagated immediately. We already have to perform a STW synchronization when increasing the minor heap size.)

I've been pondering upon this as well, but it seems unclear to me how people make use of `Gc.set` and if there could be expectations of it performing fast.

Maybe to adjust on the fly various settings by doing some regular introspection? This may be far fetched, and I was under the impression that the common use case is more of a configuration done at the very beginning of the program.

I think having it trigger a STW like for the change of the minor heap size would be fine.

**Comment by gasche** [comment]
I don't think this is necessary for the present PR, and this is only one change among several we may want to do to the statistics subsystem, so I would recommend merging the current good state now and keeping this question in mind for later.

File: runtime/gc_ctrl.c (hunk 5 above)
Comment (Octachron) [reply runtime/gc_ctrl.c:174]: This change looks like it could be done in a separate PR, isn't it?

Comment (abbysmal) [reply runtime/gc_ctrl.c:174]: I think this would indeed make sense in another PR.

**Comment by Octachron** [comment]
Cherry-picked on 5.0 as 215ed7b76 .
